### ISSUE TITLE
Race condition in CF template for custom domain mapping

### ISSUE THREAD

**Issue by chkothe** [aws/chalice#1735]
There appears to be a race condition in the CF templates generated by `chalice package` when the `api_gateway_custom_domain` feature is used, which can cause the deployment to fail spuriously. 

**Issue**
The cause is that the `ApiCustomDomainMapping` resource can be created before the respective API stage is created, _even though_ it has a reference to `RestAPI.Stage` and therefore waits at least until the API has been created. 

However, it turns out that the stage can be created long after `RestAPI` has been marked "CREATE_COMPLETE" as the below screenshot shows (the stage is `RestAPIv1Stage`). And since the domain mapping has no explicit dependency on the `RestAPIv1Stage`, it can (and sometimes will) be created immediately after `RestAPI`, which we managed to reproduce several times in a row in our API stack. 

![CF-dependencies](https://user-images.githubusercontent.com/5318120/118346876-9c768e00-b4f3-11eb-8cfc-147ec4c265d3.png)

The stack in the above screenshot was generated with a patched version of chalice, and therefore the domain mapping is created only after the `RestAPIv1Stage`. Without the patch, it is created right after the `RestAPI`, with the usual consequences:

![reproduction](https://user-images.githubusercontent.com/5318120/118348195-84a40780-b4fd-11eb-9696-5ee78f143d13.png)

**Fix**
This appears to be a long-standing ('17) issue with the SAM transforms, and I'm adding a companion PR that implements a fairly simple fix using DependsOn from [this](https://github.com/aws/serverless-application-model/issues/192#issuecomment-406834837)  post (in the thread about the issue). The most [liked](https://github.com/aws/serverless-application-model/issues/192#issuecomment-520893111) solution in that same thread obviously doesn't always work, as the screenshot shows. The problem is acknowledged also [here](https://github.com/aws/serverless-application-model/issues/783#issuecomment-467562151) in the feature issue for custom domain name support in SAM. Note that this is may also be the root cause for Chalice #1557 (with terraform in the loop).

The name of the stage resource that's created by the Serverless transform based on this SAM [documentation](https://github.com/aws/serverless-application-model/blob/master/docs/internals/generated_resources.rst#awsserverlessapi), so one would hope that the naming convention will remain stable. In theory there could be a lower-level fix in the SAM transform itself (perhaps merging that `DependsOn` clause into anything that references `RestAPI.Stage`), but we'd rather not be blocked waiting on that.

**Comment by chkothe** [comment]
Btw I cleaned up the issue description above, should now be much clearer

**Comment by MasterNayru** [comment]
@chkothe Did you test this change on a stack that had been created with the current behaviour and then try applying an update after your change had been applied? I have hit this exact same issue when dealing with raw CloudFormation templates (not SAM) but I would have thought that creating an explicit stage with the same name would cause stack updates to fail if they go from the old behaviour to this behaviour (with the stage in it)

**Comment by chkothe** [comment]
Hey, yes, we did (going from chalice 1.23 to this PR).

So if you apply this fix to an existing stack, what it does is that it replaces a literal string (in a base path mapping resource) by a reference to a stage resource, which in turn resolves to the same string as before. 

And the stage resource that's referenced has already been an implicit part of the SAM template all along, so no physical resource gets added, removed, or updated by this. Actually, when you apply the change, the deployed base path mapping wouldn’t even be updated either, since nothing in the physical resource actually changed (same string as before) -- the only thing that changes is the dependency order that CFN is required to adhere to. 

Therefore, if I was worried about anything, it’d be that CFN gives its trademark 'error' that “nothing actually changed” when doing a deploy with the fix applied. Now in the context of chalice, the version update would incur some other material changes in the stack (namely the lambda code package or the dependencies layer image), so that particular hiccup wouldn't happen.

**Comment by MasterNayru** [comment]
@chkothe Right, so there is a Stage resource being created but it's just not being made an explicit dependency. Looks like a pretty straightforward improvement to me, then.

The issue I was referring to happens in the case where you create a deployment with a string name but don't have a Stage resource in the stack, but then later try to use a Stage resource with the same name as whatever was set before. It seems like the SAM resource is doing the right thing in that regard, though.

**Comment by jonathanabila** [comment]
@chkothe thank you for the solution! <3

Any updates on the revision of this ticket? This was a serious problem for me and after using the code in PR everything worked as expected.

It would be awesome if we could include this fix in the new release - it's small and probably quick to review.

**Comment by chkothe** [comment]
@jonathanabila sure np. That issue is most likely unrelated (the PR doesn't add or rename any resources), e.g., could be a config change while iterating on the deployment.

**Comment by jonathanabila** [comment]
> @jonathanabila sure np. That issue is most likely unrelated (the PR doesn't add or rename any resources), e.g., could be a config change while iterating on the deployment.

Ok - I'll try again in another environment - thanks again!

**Comment by jonathanabila** [comment]
> @jonathanabila sure np. That issue is most likely unrelated (the PR doesn't add or rename any resources), e.g., could be a config change while iterating on the deployment.

Hi @chkothe I did another deployment to create a new environment and got the same error:
![image](https://user-images.githubusercontent.com/31670446/130288466-fde19669-a613-486a-b397-bc9700298c54.png)

Do you have any information I can give you to help with the debugging process?

**Comment by chkothe** [comment]
Hi Johnathan, if you want to be 120% sure that it's not caused by this PR, you might also try a different way that comes down to the same end result (delaying the creation of the custom domain mapping until after the stage has been created):
https://github.com/chkothe/chalice/commit/b464c7f9b3826494ae7ac63ce7463bced20f546e

... that's the commit for an alternative PR I had made with the same goal (namely fixing the resource creation order), which works by adding a DependsOn clause to the CF template instead of a cross-reference. Again, none of them *should* affect naming in any way, only the order in which resources are created, but surely there's a lot going on at deploy time and the latter definitely makes it more clear what's changing (although unfortunately it's not the prettiest depends-on due to the way SAM names its resources). So if you were to use that, and yet you *still* get mystery stages appearing every now and then in your CF stack then it'd prove pretty conclusively that it's caused by some other glitch (whatever that may be).

**Comment by Nath-P** [comment]
Is there a reason [chkothe](https://github.com/chkothe) 's commit can't  be merged?  I hit this issue today.  (deploying using cdk)    I can work around by deploying the stack first without the custom mapping and then adding it later but its a pain

**Comment by AlexFromXD** [comment]
any update of this ?

**Comment by Ankushpandey-ti** [comment]
any update on this?

**Comment by isaacsancheza** [comment]
A work around when using CDK.

```python3
from typing import Any, cast

import aws_cdk as cdk
from aws_cdk import aws_apigateway as apigatewayv1
from aws_cdk import aws_apigatewayv2 as apigateway
from constructs import Construct
from chalice.cdk import Chalice


class ChaliceNestedStack(cdk.NestedStack):
    def __init__(
            self, 
            scope: Construct, 
            construct_id: str, 
            /, 
            *, 
            source_dir: str = 'chalice',
            stage_config: dict[str, Any],
        ) -> None:
        super().__init__(scope, construct_id)
        
        self.app = Chalice(
            self, 
            'App', 
            source_dir=source_dir,
            stage_config=stage_config,
        )

        rest_api = cast(apigateway.CfnApi, self.app.get_resource('RestAPI'))
        custom_domain = cast(apigateway.CfnDomainName, self.app.get_resource('ApiGatewayCustomDomain'))
        base_path_mapping = cast(apigatewayv1.CfnBasePathMapping, self.app.get_resource('ApiGatewayCustomDomainMapping'))

        custom_domain.add_dependency(rest_api)        
        base_path_mapping.add_dependency(custom_domain)

```

**Comment by empeekdev** [comment]
Almost 3 years and no fix?
